### PR TITLE
multisig-test-vectors

### PR DESCRIPTION
**Pull request by danielradu10**

### CODE REVIEW

File: signing/mcl/multisig/jsonGenerator.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+			AggregatedSignature: hex.EncodeToString(aggregatedSig),
+			ErrorMessage:        predefinedTest.ExpectedError,
+			TestName:            predefinedTest.TestName,
+		})
+		sigPairs = sigPairs[:0]
+	}
+
+	jsonFileContent.TestVectors = testVectors
+
+	b, _ := json.MarshalIndent(jsonFileContent, "", " ")
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply signing/mcl/multisig/jsonGenerator.go:186]: and perhaps rename to marshalledJsonFileContent

File: signing/mcl/multisig/vectors_test.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	"github.com/multiversx/mx-chain-core-go/hashing/blake2b"
+	"github.com/stretchr/testify/require"
+
+	crypto "github.com/multiversx/mx-chain-crypto-go"
+	"github.com/multiversx/mx-chain-crypto-go/signing"
+	"github.com/multiversx/mx-chain-crypto-go/signing/mcl"
+)
+
+// TestVector defines the data structure used to unmarshal the JSON file
+type TestVector struct {
```
Comment (sstanculeanu) [inline signing/mcl/multisig/vectors_test.go:19]: why not use the already defined structure from jsonGenerator.go?

File: signing/mcl/multisig/vectors_test.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	suite := mcl.NewSuiteBLS12()
+	testVectors, err := createTestSetup(suite, "multisigKOSKAggSig.json")
+	require.Nil(t, err)
+
+	for _, testVector := range testVectors {
+		testName := testVector.testName
+		signatures := testVector.signatures
+		pubKeys := testVector.publicKeys
+		expectedError := testVector.expectedError
+		aggregatedSig := testVector.aggregatedSig
```
Comment (sstanculeanu) [inline signing/mcl/multisig/vectors_test.go:45]: new vars not necessarily needed

valid for all tests

File: signing/mcl/multisig/vectors_test.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+}
+
+func TestAggregateSignaturesNonKOSK(t *testing.T) {
+	t.Parallel()
+
+	lls := &BlsMultiSigner{}
+	hasher, err := blake2b.NewBlake2bWithSize(blsHashSize)
+	require.Nil(t, err)
+	lls.Hasher = hasher
```
Comment (sstanculeanu) [inline signing/mcl/multisig/vectors_test.go:89]:
```suggestion
	hasher, err := blake2b.NewBlake2bWithSize(blsHashSize)
	require.Nil(t, err)
	
	lls := &BlsMultiSigner{
		Hasher: hasher,
	}
```

File: signing/mcl/multisig/vectors_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+}
+
+func TestVerifyAggregatedSigNonKOSK(t *testing.T) {
+	t.Parallel()
+
+	lls := &BlsMultiSigner{}
+	hasher, err := blake2b.NewBlake2bWithSize(blsHashSize)
+	require.Nil(t, err)
+	lls.Hasher = hasher
```
Comment (sstanculeanu) [inline signing/mcl/multisig/vectors_test.go:119]:
```suggestion
	hasher, err := blake2b.NewBlake2bWithSize(blsHashSize)
	require.Nil(t, err)
	
	lls := &BlsMultiSigner{
		Hasher: hasher,
	}
```

File: signing/mcl/multisig/vectors_test.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		aggregatedSig := testVector.AggregatedSignature
+
+		message := testVector.Message
+		expectedError := errors.New(testVector.ErrorMessage)
+		if testVector.ErrorMessage == "noError" {
+			expectedError = nil
+		}
+
+		var pubKeys []crypto.PublicKey
+		var sigs [][]byte
```
Comment (sstanculeanu) [inline signing/mcl/multisig/vectors_test.go:169]:
```suggestion
		pubKeys := make([]crypto.PublicKey, 0, len(signatures))
		sigs := make([][]byte, 0, len(signatures))
```

File: signing/mcl/multisig/jsonGenerator.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	"os"
+
+	crypto "github.com/multiversx/mx-chain-crypto-go"
+	"github.com/multiversx/mx-chain-crypto-go/signing"
+	"github.com/multiversx/mx-chain-crypto-go/signing/mcl"
+)
+
+const blsHashSize = 16
+
+//"message to be signed or to be verified is nil"
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:15]: removed

File: signing/mcl/multisig/jsonGenerator.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+// Key defines a tuple of public key and private key
+type Key struct {
+	PubKey     crypto.PublicKey
+	PrivateKey crypto.PrivateKey
+}
+
+func createKeyPairs(grSize uint16, suite crypto.Suite) []Key {
+	kg := signing.NewKeyGenerator(suite)
+	var keys []Key
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:95]: updated

File: signing/mcl/multisig/jsonGenerator.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+// generateJSONFileNonKOSKForAggregateSignaturesTests for NonKOSK VerifyAggregatedSigTests
+func generateJSONFileNonKOSKForVerifyAggregatedSig(signer crypto.LowLevelSignerBLS) error {
+	return generateJSONFile(signer, predefinedNonKOSKVerifyAggregatedSigTests, "multisigNonKOSKVerifyAggSig.json")
+}
+
+// generateJSONFile generates the JSON file for KOSK/NonKOSK, should be used only once
+func generateJSONFile(lls crypto.LowLevelSignerBLS, predefinedTests []PredefinedTest, filename string) error {
+	suite := mcl.NewSuiteBLS12()
+
+	mapKeys := createKeyPairs(uint16(5), suite)
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:131]: updated

File: signing/mcl/multisig/jsonGenerator.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+	var keyPairs []KeyPair
+	var testVectors []TestVectorElement
+	var jsonFileContent JSONFileContent
+
+	var sigShares [][]byte
+	var pubKeys []crypto.PublicKey
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:138]: updated

File: signing/mcl/multisig/jsonGenerator.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+			PublicKey:  hex.EncodeToString(pk),
+			PrivateKey: hex.EncodeToString(sk),
+		})
+
+		pubKeys = append(pubKeys, key.PubKey)
+	}
+
+	jsonFileContent.Keys = keyPairs
+
+	var sigPairs []SignaturePair
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:154]: updated

File: signing/mcl/multisig/jsonGenerator.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+			sigShares = append(sigShares, sig)
+			sigPairs = append(sigPairs, SignaturePair{
+				Signature: hex.EncodeToString(sig),
+				PublicKey: hex.EncodeToString(pk),
+			})
+		}
+
+		aggregatedSig, _ := lls.AggregateSignatures(suite, sigShares, pubKeys)
+		sigShares = sigShares[:0]
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:172]: here is why I thought it should be used: for every message from the predefinedTests we will have a new set of sigShares, and I want to append to a new, empty slice. I found that sigShares[:0] removes all the content from a slice so I use it everytime I finish with AggregateSignatures

File: signing/mcl/multisig/jsonGenerator.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+		testVectors = append(testVectors, TestVectorElement{
+			Signatures:          sigPairs,
+			Message:             hex.EncodeToString([]byte(predefinedTest.Message)),
+			AggregatedSignature: hex.EncodeToString(aggregatedSig),
+			ErrorMessage:        predefinedTest.ExpectedError,
+			TestName:            predefinedTest.TestName,
+		})
+		sigPairs = sigPairs[:0]
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:181]: same with sigShares

File: signing/mcl/multisig/jsonGenerator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:186]: updated

File: signing/mcl/multisig/jsonGenerator.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	jsonFileContent.TestVectors = testVectors
+
+	b, _ := json.MarshalIndent(jsonFileContent, "", " ")
+
+	err := os.WriteFile(filename, b, 0644)
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply signing/mcl/multisig/jsonGenerator.go:188]: updated

File: signing/mcl/multisig/vectors_test.go (hunk 3 above)
Comment (danielradu10) [reply signing/mcl/multisig/vectors_test.go:45]: updated

File: signing/mcl/multisig/vectors_test.go (hunk 4 above)
Comment (danielradu10) [reply signing/mcl/multisig/vectors_test.go:89]: updated

File: signing/mcl/multisig/vectors_test.go (hunk 5 above)
Comment (danielradu10) [reply signing/mcl/multisig/vectors_test.go:119]: updated

File: signing/mcl/multisig/vectors_test.go (hunk 6 above)
Comment (danielradu10) [reply signing/mcl/multisig/vectors_test.go:169]: updated

File: signing/mcl/multisig/testData/data.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+// PredefinedTest defines the data used for testing
+type PredefinedTest struct {
+	TestName      string
+	Message       string
+	ExpectedError string
+}
+
+// predefinedKOSKAggregateSignaturesTests defines the scenarios for testing the AggregateSignatures for KOSK
+var predefinedKOSKAggregateSignaturesTests = []PredefinedTest{
```
Comment (AdoAdoAdo) [inline signing/mcl/multisig/testData/data.go:15]: these structures at lines 15, 24, 33, 42 are very similar (duplicated). We can keep only one definition (rename it to something more generic) and use it for the different test vectors.

File: signing/mcl/multisig/vectors_test.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	var jsonContent testData2.JSONFileContent
+	err = json.Unmarshal(content, &jsonContent)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, testVector := range jsonContent.TestVectors {
+
+		testName := testVector.TestName
```
Comment (AdoAdoAdo) [inline signing/mcl/multisig/vectors_test.go:149]: can you extract the lines 149 -> 175 into a function that creates a TestVector instance?

Comment (danielradu10) [reply signing/mcl/multisig/vectors_test.go]: updated

**Review by AdoAdoAdo — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Pull request merged**